cli/ballerina-cli/src/main/java/io/ballerina/cli/diagnostics/AnnotateDiagnostics.java

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (https://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.cli.diagnostics;
+
+import io.ballerina.compiler.internal.diagnostics.StringDiagnosticProperty;
+import io.ballerina.projects.Document;
+import io.ballerina.projects.DocumentId;
+import io.ballerina.projects.Module;
+import io.ballerina.projects.ModuleName;
+import io.ballerina.projects.Package;
+import io.ballerina.projects.internal.PackageDiagnostic;
+import io.ballerina.tools.diagnostics.Diagnostic;
+import io.ballerina.tools.diagnostics.DiagnosticInfo;
+import io.ballerina.tools.diagnostics.DiagnosticSeverity;
+import io.ballerina.tools.diagnostics.Location;
+import io.ballerina.tools.text.LinePosition;
+import io.ballerina.tools.text.TextDocument;
+import org.jline.jansi.Ansi;
+import org.jline.terminal.TerminalBuilder;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static io.ballerina.cli.diagnostics.DiagnosticAnnotation.NEW_LINE;
+import static io.ballerina.cli.diagnostics.DiagnosticAnnotation.SEVERITY_COLORS;
+import static io.ballerina.cli.diagnostics.DiagnosticAnnotation.getColoredString;
+import static io.ballerina.cli.utils.OsUtils.isWindows;
+
+/**
+ * This class is used to generate diagnostic annotated messages from diagnostics.
+ *
+ * @since 2201.9.0
+ */
+public class AnnotateDiagnostics {
+
+    private static final String COMPILER_ERROR_PREFIX = "BCE";
+    private static final int SYNTAX_ERROR_CODE_THRESHOLD = 1000;
+    private static final int MISSING_TOKEN_KEYWORD_CODE_THRESHOLD = 400;
+    private static final int INVALID_TOKEN_CODE = 600;
+    private static final int NO_TRUNCATE_WIDTH = 999;
+
+    private final Map<String, Document> documentMap;
+    private int terminalWidth;
```

Comment:
I think we can make this final as well by checking if the value is 0 in the constructor